### PR TITLE
Client to Broker/Wrapper to Broker connection

### PR DESCRIPTION
**Pull request by mrzzzrm**

### CODE REVIEW

File: source/broker/BrokerRenderingSession.cpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+bool BrokerRenderingSession::worker()
+{
+    if (m_wrapperConnection->bytesAvailable() > 0)
+    {
+        std::cout << "Available: " << m_wrapperConnection->bytesAvailable() << std::endl;
+    }
+
+    std::this_thread::sleep_for(std::chrono::duration<float>(1.0f));
+
+    return true;
```
Comment (apopiak) [inline source/broker/BrokerRenderingSession.cpp]: never returns `false` --> zombie thread :-P

File: source/broker/BrokerServer.cpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    json["SharedMemoryWriteName"] = shmDevice->readName();
+    json["SharedMemorySize"] = (int)shmDevice->sharedMemorySize();
+
+    auto msg = QJsonDocument(json).toJson();
+
+    clientConnection->write(common::BlobWriter::encode(msg));
+    clientConnection->flush();
+
+    auto device = std::unique_ptr<QIODevice>(std::move(shmDevice));
+    auto session = std::unique_ptr<BrokerRenderingSession>(new BrokerRenderingSession(std::move(device), clientConnection));
```
Comment (apopiak) [inline source/broker/BrokerServer.cpp]: `device` is not used

File: source/client/ClientBrokerConnection.cpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+#include <QtCore/QByteArray>
+
+ClientBrokerConnection::ClientBrokerConnection(QObject * parent):
+    QObject(parent),
+    m_port(-1)
+{
+    m_socket = new QTcpSocket(this);
+
+    m_socketStream.setDevice(m_socket);
+    m_socketStream.setVersion(QDataStream::Qt_4_0);
```
Comment (apopiak) [inline source/client/ClientBrokerConnection.cpp]: not really needed (never used AFAIK)

File: source/common/BlobReader.cpp [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    char * data;
+    unsigned int size;
+
+    stream.readBytes(data, size);
+
+    if (size == 0) {
+        return QByteArray();
+    }
+
+    m_buf.remove(0, 4 + size);
```
Comment (apopiak) [inline source/common/BlobReader.cpp]: add an explanation for the magic number?

File: source/common/RandomStringGenerator.cpp [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+         'G','H','I','J','K',
+         'L','M','N','O','P',
+         'Q','R','S','T','U',
+         'V','W','X','Y','Z',
+         'a','b','c','d','e','f',
+         'g','h','i','j','k',
+         'l','m','n','o','p',
+         'q','r','s','t','u',
+         'v','w','x','y','z'
+        };
```
Comment (apopiak) [inline source/common/RandomStringGenerator.cpp]: make it static?

File: source/common/SharedMemoryDevice.cpp [hunk 6]
```diff
@@ -106,6 +129,10 @@ qint64 SharedMemoryDevice::readData(char *data, qint64 maxlen)
 
 qint64 SharedMemoryDevice::writeData(const char *data, qint64 len)
 {
+#if VERBOSE
+    //std::cout << "SharedMemoryDevice: WritingQ " << len << " bytes" << std::endl;
```
Comment (apopiak) [inline source/common/SharedMemoryDevice.cpp]: unnecessary outcommenting

File: source/common/SharedMemoryDevice.cpp [hunk 7]
```diff
@@ -170,6 +208,13 @@ bool SharedMemoryDevice::setUpReading()
     m_readQueue.reset(new ProducerConsumerQueue());
     m_readMem.reset(new SharedMemory(m_readName, m_size));
     bool success = m_readMem->attach();
+    if (m_type == Type_Server)
+    {
+        // initialize shared memory header
+        auto memLock = m_readMem->lock();
+        *m_readMem->isOpenPtr() = true;
+        *m_readMem->lengthPtr() = 0;
+    }
```
Comment (apopiak) [inline source/common/SharedMemoryDevice.cpp]: extract to method (the same code is used twice for `m_readMem`and `m_writeMem`

File: source/common/SharedMemoryDevice.cpp [hunk 8]
```diff
@@ -170,6 +208,13 @@ bool SharedMemoryDevice::setUpReading()
     m_readQueue.reset(new ProducerConsumerQueue());
     m_readMem.reset(new SharedMemory(m_readName, m_size));
     bool success = m_readMem->attach();
+    if (m_type == Type_Server)
```
Comment (apopiak) [inline source/common/SharedMemoryDevice.cpp]: check for success
(I forgot that in my PR ^^)

File: source/common/SharedMemoryDevice.h [hunk 9]
```diff
@@ -18,13 +18,30 @@ class SharedMemoryDevice : public QIODevice
     Q_OBJECT
 
 public:
-    SharedMemoryDevice(const QString & readName, const QString &writeName, size_t size = 1024 * 1024);
+    enum ClientServerType
+    {
+        Type_Server,
+        Type_Client
+    };
```
Comment (apopiak) [inline source/common/SharedMemoryDevice.h]: Why not use:
```
enum class Type
{
    Client,
    Server
}
```
?

Comment (apopiak) [inline source/common/SharedMemoryDevice.h]: Also: it seems more like a boolean flag `initializeHeaders`

File: source/broker/BrokerServer.cpp (hunk 2 above)
Comment (mrzzzrm) [inline source/broker/BrokerServer.cpp]: would be too long in one line imo, since we're casting here

File: source/sandboxes/CMakeLists.txt [hunk 10]
```diff
@@ -4,3 +4,6 @@ if (NOT OPTION_BUILD_SANDBOXES)
 endif()
 
 add_subdirectory(IPC)
+add_subdirectory(JSONSandbox)
```
Comment (apopiak) [inline source/sandboxes/CMakeLists.txt]: needed?

File: source/wrapper/lib/trace/SharedMemoryOutStream.cpp [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+SharedMemoryOutStream::~SharedMemoryOutStream()
+{
+
+}
+
+bool SharedMemoryOutStream::write(const void *buffer, size_t length)
+{
+    m_shmDevice->write((const char *)buffer, length);
```
Comment (apopiak) [inline source/wrapper/lib/trace/SharedMemoryOutStream.cpp]: c cast, oh neeeeein

**Comment by apopiak** [comment]
LGTM

File: source/wrapper/lib/trace/SharedMemoryOutStream.cpp (hunk 11 above)
Comment (mrzzzrm) [inline source/wrapper/lib/trace/SharedMemoryOutStream.cpp]: clean, simple, easy to read, every one is used to it
vs the
elitist c++ cancer that is static_oh_maybe_reinterpret_cast<>()

**Pull request merged**